addon/record-data.js

```diff
@@ -0,0 +1,989 @@
+ *
+ * @class M3SchemaInterface
+ */
+class M3SchemaInterface {
+  /**
+   * @param {M3RecordData} recordData
+   */
+  constructor(recordData) {
+    this.recordData = recordData;
+    this._keyBeingResolved = null;
+    this._refKeyDepkeyMap = {};
+    this._suppressNotifications = false;
+  }
+
+  /**
+   * @param {string} key
+   * @private
+   */
+  _beginDependentKeyResolution(key) {
+    assert(
+      'Do not invoke `SchemaInterface` method `_beginDependentKeyResolution` without ending the resolution of previous key.',
+      this._keyBeingResolved === null
+    );
+    this._keyBeingResolved = key;
+  }
+
+  /**
+   * @param {string} key
+   * @private
+   */
+  _endDependentKeyResolution(key) {
+    assert(
+      'Do not invoke `SchemaInterface` method `_endDependentKeyResolution` without begining the resolution of the key.',
+      key && this._keyBeingResolved === key
+    );
+    this._keyBeingResolved = null;
+  }
+
+  _getDependentResolvedKeys(refKey) {
+    return this._refKeyDepkeyMap[refKey];
+  }
+
+  /**
+   * Get the attribute name from the model.
+   * This can be useful if your payload keys are different from your attribute names;
+   * e.g. if your api adds a prefix to attributes that should be interpreted as references.
+   *
+   * @param {string} name name of the attribute
+   * @returns {Object} value of the attribute
+   */
+  getAttr(name) {
+    let value = this.recordData.getAttr(name);
+    const keyBeingResolved = this._keyBeingResolved;
+    assert(
+      'Do not manually call methods on `schemaInterface` outside of schema resolution hooks such as `computeAttributeReference`',
+      keyBeingResolved
+    );
+
+    // it might be nice to avoid doing this if we know that we don't need to
+    // support depkeys from the server changing between requests
+    this._refKeyDepkeyMap[name] = this._refKeyDepkeyMap[name] || [];
+    let refKeyMap = this._refKeyDepkeyMap[name];
+    if (refKeyMap.indexOf(keyBeingResolved) < 0) {
+      refKeyMap.push(this._keyBeingResolved);
+    }
+
+    return value;
+  }
+
+  /**
+   * Set attribute for the recordData
+   *
+   * @param {string} key
+   * @param {Object} value
+   */
+  setAttr(key, value) {
+    this.recordData.setAttr(key, value, this._suppressNotifications);
+  }
+}
+
+export default class M3RecordData {
+  /**
+   * @param {string} modelName
+   * @param {string} id
+   * @param {number} [clientId]
+   * @param {DS.Store} storeWrapper
+   * @param {SchemaManager} schemaManager
+   * @param {M3RecordData} [parentRecordData]
+   * @param {M3RecordData} [baseRecordData]
+   */
+  constructor(
+    modelName,
+    id,
+    clientId,
+    storeWrapper,
+    schemaManager,
+    parentRecordData,
+    baseRecordData
+  ) {
+    this.modelName = modelName;
+    this.clientId = clientId;
+    this.id = id;
+    this.storeWrapper = storeWrapper;
+
+    if (!storeWrapper.recordDataFor) {
+      storeWrapper.recordDataFor = storeWrapper.modelDataFor;
+    }
```

Comment:
rather than mutate the `storeWrapper` we get from ember-data let's wrap access to this in a helper (similar to `recordDataFor` and swap out its implementation at build time based on the version of ember data

Reply:
done